File: renew_therose.py
#!/usr/bin/env python3

import os,re,sys,time,requests
from seleniumbase import SB

# 环境变量
EMAIL = os.environ.get("EMAIL") or ""            # 邮箱
PASSWORD = os.environ.get("PASSWORD") or ""      # 密码
TG_BOT_TOKEN = os.environ.get("TG_BOT_TOKEN") or ""  # tg通知 bot token
TG_CHAT_ID = os.environ.get("TG_CHAT_ID") or ""      # tg通知 chat_id id

# 代理配置（由 setup_proxy.sh 写入 GITHUB_ENV）
IS_PROXY = os.environ.get("IS_PROXY", "false").lower() == "true"
PROXY_SERVER = os.environ.get("PROXY_SERVER", "").strip() or "socks5://127.0.0.1:1080"

BASE_URL = "https://client.therose.cloud/login"

# 获取当前出口ip（用于确认代理是否生效）
def get_current_ip(proxy_server: str = "") -> str:
    proxies = None
    if proxy_server:
        proxies = {"http": proxy_server, "https": proxy_server}
    response = requests.get("https://api.ip.sb/ip", proxies=proxies, timeout=15)
    response.raise_for_status()
    return response.text.strip()

# 检查必要变量
if not EMAIL or not PASSWORD:
    print("❌ 请设置环境变量 EMAIL 和 PASSWORD")
    sys.exit(1)

# 点击续期按钮
def click_extend_button(sb):
    selectors = [
        'span:contains("Extend")',
        'button:contains(title="Extend")',
    ]
    for sel in selectors:
        try:
            if sb.find_element(sel, timeout=2):
                print(f"✅ 找到按钮，选择器: {sel}")
                sb.uc_click(sel, timeout=5)
                print("✅ 点击成功")
                return True, {}
        except:
            continue
    try:
        btn = sb.find_element('button:contains("Extend")', timeout=2)
        sb.driver.execute_script("arguments[0].click();", btn)
        print("✅ 通过 JavaScript 点击成功")
        return True, {}
    except Exception as e:
        return False, {"error": str(e)}

# 检查续期是否成功
def check_renewal_success(sb):
    """检查是否出现续期成功的提示"""
    success_selectors = [
        '.alert-success',
        '.alert.alert-success',
        'div[role="alert"].alert-success',
        'div.alert-success',
        'span:contains("successfully purchased")',
        'div:contains("successfully purchased")'
    ]

    print("⏳ 等待5秒检查续期结果...")
    time.sleep(5)

    for selector in success_selectors:
        try:
            element = sb.find_element(selector, timeout=2)
            if element:
                text = element.text
                print(f"✅ 发现成功提示！选择器: {selector}")
                print(f"📝 提示内容: {text}")
                return True, text
        except:
            continue

    try:
        page_source = sb.get_page_source()
        if "successfully purchased" in page_source.lower():
            print("✅ 页面源码中发现 'successfully purchased' 关键词")
            return True, "服务器已成功续期"
    except:
        pass

    return False, "未检测到续期成功提示"

# 掩码邮箱（只用于通知展示，脱敏）
def mask_email(email: str) -> str:
    if '@' in email:
        name, domain = email.split('@', 1)
        if len(name) > 4:
            return f"{name[:2]}****{name[-2:]}@{domain}"
        return f"{name}@{domain}"
    return email[:2] + '****' if email else "（未配置）"

# 通知格式
def format_notification(status: str, extra: str = "", error: str = "") -> str:
    local_time = time.gmtime(time.time() + 8 * 3600)
    now = time.strftime("%Y-%m-%d %H:%M:%S", local_time)
    lines = [
        "🌹 The Rose Cloud 续期通知",
        "",
        f"{status}",
        f"👤 登录账户: {mask_email(EMAIL)}",
    ]
    if extra:
        lines.append(extra)
    if error:
        lines.append(f"⚠️ 错误信息: {error}")
    lines.append(f"⏱️ 执行时间: {now}")
    return "\n".join(lines)

# 发送tg通知
def send_tg(token, chat_id, message):
    if not token or not chat_id:
        return
    url = f"https://api.telegram.org/bot{token}/sendMessage"
    try:
        resp = requests.post(url, json={"chat_id": chat_id, "text": message}, timeout=10)
        if resp.status_code == 200:
            print("📨 Telegram 通知已发送")
        else:
            print(f"❌ Telegram 发送失败: {resp.text}")
    except Exception as e:
        print(f"❌ Telegram 发送异常: {e}")

# 等待 Turnstile 验证通过（只认真实 token，避免文案误判）
def wait_for_turnstile_pass(sb, timeout=30):
    pending_markers = [
        "verify you are human",
        "确认您是真人",
        "just a moment",
        "checking your browser",
        "正在验证",
        "验证中",
        "stuck?",
        "卡住",
    ]
    start = time.time()
    while time.time() - start < timeout:
        try:
            token = sb.execute_script(
                "var el=document.querySelector("
                "'[name=\"cf-turnstile-response\"], textarea[name=\"cf-turnstile-response\"], "
                "input[name=\"cf-turnstile-response\"]');"
                "return el ? (el.value || '') : '';"
            )
            if token and len(token) > 20:
                print("✅ Turnstile token 已就绪")
                return True
        except Exception:
            pass
        try:
            page_lower = sb.get_page_source().lower()
            if any(x in page_lower for x in pending_markers):
                # 仍在验证中，继续等
                sb.sleep(1)
                continue
        except Exception:
            pass
        sb.sleep(1)
    print("❌ Turnstile 验证超时未通过（未拿到 token）")
    return False

# 采集登录页错误信息
def extract_login_error(sb) -> str:
    selectors = [
        ".alert-danger",
        ".alert.alert-danger",
        "div[role='alert']",
        ".invalid-feedback",
        ".form-error",
        ".error-message",
        "div.alert",
    ]
    for sel in selectors:
        try:
            if sb.is_element_visible(sel):
                text = (sb.get_text(sel) or "").strip()
                if text:
                    return text[:300]
        except Exception:
            continue
    try:
        body = (sb.get_text("body") or "")[:500]
        for kw in ["invalid", "incorrect", "failed", "error", "wrong", "captcha", "验证", "错误", "失败"]:
            if kw in body.lower():
                return body.replace("\n", " ").strip()[:200]
    except Exception:
        pass
    return ""

# 登录流程
def login(sb, email, password):
    print("🌐 打开登录页面...")
    # UC 模式：打开后短暂断开 chromedriver，降低 CF 检测
    sb.uc_open_with_reconnect(BASE_URL, reconnect_time=4)
    sb.wait_for_ready_state_complete()
    sb.sleep(2)

    def fill_credentials():
        print("📧 填写邮箱...")
        sb.type("#login_form_email", email, timeout=10)
        print("🔑 填写密码...")
        sb.type("#login_form_password", password, timeout=10)
        sb.sleep(0.5)
        # 确认已写入（非 placeholder）
        try:
            v = sb.get_value("#login_form_email") or ""
            if email not in v and v != email:
                print(f"⚠️ 邮箱字段异常: {v!r}，重试填写")
                sb.clear("#login_form_email")
                sb.type("#login_form_email", email, timeout=10)
                sb.clear("#login_form_password")
                sb.type("#login_form_password", password, timeout=10)
        except Exception as e:
            print(f"⚠️ 校验表单字段失败: {e}")

    fill_credentials()

    # 处理 Turnstile：最多重试 3 次，通过后再点登录
    print("🛡 处理 Turnstile...")
    turnstile_passed = False
    for attempt in range(1, 4):
        # 每次尝试前确保凭证仍在
        try:
            v = sb.get_value("#login_form_email") or ""
            if not v or email not in v:
                fill_credentials()
        except Exception:
            fill_credentials()

        try:
            sb.uc_gui_click_captcha(retry=True, blind=True)
            print(f"✅ 第 {attempt} 次点击 Turnstile")
            sb.sleep(5)
        except Exception as e:
            print(f"⚠️ uc_gui_click_captcha 异常 (第{attempt}次): {e}")

        if wait_for_turnstile_pass(sb, timeout=25):
            turnstile_passed = True
            break
        print(f"⏳ 第 {attempt} 次未通过，重试...")
        sb.sleep(2)

    if not turnstile_passed:
        err = extract_login_error(sb) or "Turnstile 验证未通过"
        print(f"❌ {err}")
        sb.save_screenshot("login_failed.png")
        return False, f"{sb.get_current_url()} | {err}"

    # 点击前再确认凭证
    try:
        v = sb.get_value("#login_form_email") or ""
        if not v or email not in v:
            fill_credentials()
    except Exception:
        fill_credentials()

    print("🔑 点击登录按钮...")
    try:
        sb.uc_click('button:contains("Sign in")')
    except Exception as e:
        print(f"⚠️ uc_click 失败，尝试 JS 点击: {e}")
        try:
            btn = sb.find_element('button:contains("Sign in")', timeout=5)
            sb.driver.execute_script("arguments[0].click();", btn)
        except Exception as e2:
            sb.save_screenshot("login_failed.png")
            return False, f"点击 Sign in 失败: {e2}"

    sb.sleep(4)

    # 提交后若再次出现验证码，再点一次
    try:
        page_lower = sb.get_page_source().lower()
        if any(x in page_lower for x in ["正在验证", "verify you are human", "cf-turnstile"]):
            print("🛡 登录后再次出现 Turnstile，继续处理...")
            sb.uc_gui_click_captcha(retry=True, blind=True)
            wait_for_turnstile_pass(sb, timeout=20)
            sb.sleep(2)
    except Exception:
        pass

    for i in range(30):
        current_url = sb.get_current_url()
        page_title = sb.get_title() or ""
        print(f"📄 [{i+1}/30] URL: {current_url} | Title: {page_title}")
        if "panel" in current_url or "/dashboard" in current_url:
            print("✅ 登录成功，已跳转到 Dashboard")
            return True, current_url
        if i >= 3 and i % 5 == 0:
            err = extract_login_error(sb)
            if err and any(k in err.lower() for k in ["invalid", "incorrect", "wrong", "错误", "失败"]):
                print(f"❌ 检测到登录错误: {err}")
                sb.save_screenshot("login_failed.png")
                return False, f"{current_url} | {err}"
        time.sleep(1)

    err = extract_login_error(sb) or "超时未跳转 Dashboard"
    print(f"❌ 登录失败: {err} | URL: {sb.get_current_url()}")
    sb.save_screenshot("login_failed.png")
    return False, f"{sb.get_current_url()} | {err}"

# 主流程
def main():
    print("🚀 启动浏览器")

    sb_kwargs = {"uc": True, "headless": False}

    if IS_PROXY:
        print(f"🔗 挂载代理: {PROXY_SERVER}")
        sb_kwargs["proxy"] = PROXY_SERVER
    else:
        print("🍭 未使用代理，直连访问")

    with SB(**sb_kwargs) as sb:
        ip = ""
        try:
            ip = get_current_ip(PROXY_SERVER if IS_PROXY else "")
            print(f"📍 当前出口IP: {ip}")
        except Exception as e:
            print(f"⚠️ 获取出口 IP 失败: {e}")

        success, url = login(sb, EMAIL, PASSWORD)

        if not success:
            msg = format_notification("❌ 登录失败", error=str(url))
            print(msg)
            send_tg(TG_BOT_TOKEN, TG_CHAT_ID, msg)
            return

        print("📄 开始续期流程...")

        ok, info = click_extend_button(sb)
        if not ok:
            msg = format_notification("❌ 续期失败", error=f"点击 Extend 按钮失败: {info.get('error')}")
            print(msg)
            send_tg(TG_BOT_TOKEN, TG_CHAT_ID, msg)
            return

        time.sleep(1)

        try:
            button = sb.find_element('button:contains("Order now")', timeout=5)
            if button:
                print("🛒 点击 Order now 按钮...")
                sb.uc_click('button:contains("Order now")')
                print("✅ 已点击 Order now 按钮")
            else:
                msg = format_notification("❌ 续期失败", error="未找到 Order now 按钮")
                print(msg)
                send_tg(TG_BOT_TOKEN, TG_CHAT_ID, msg)
                return
        except Exception as e:
            msg = format_notification("❌ 续期失败", error=f"点击 Order now 失败: {e}")
            print(msg)
            send_tg(TG_BOT_TOKEN, TG_CHAT_ID, msg)
            return

        print("🔍 检查续期结果...")
        renewal_success, renewal_msg = check_renewal_success(sb)

        ip_extra = f"🌐 出口IP: {ip}" if ip else ""

        if renewal_success:
            status = "✅ 续期成功"
            extra = renewal_msg
            print(f"✅ 续期成功！{renewal_msg}")
            sb.save_screenshot("renewal_success.png")
        else:
            status = "❌ 续期可能失败"
            extra = "请登录后台检查"
            print(f"❌ 续期可能失败: {renewal_msg}")
            sb.save_screenshot("renewal_failed.png")

        msg = format_notification(
            status,
            extra=" | ".join(filter(None, [extra, ip_extra])),
            error="" if renewal_success else renewal_msg,
        )
        print(msg)
        send_tg(TG_BOT_TOKEN, TG_CHAT_ID, msg)

    print("🏁 脚本执行完毕")

if __name__ == "__main__":
    main()

File: renew_therose.js
/**
 * The Rose Cloud 自动续期（puppeteer-real-browser 版）
 *
 * 过 Cloudflare Turnstile 的核心：puppeteer-real-browser 的 turnstile:true 自动求解，
 * 且其反检测指纹足够真，通常让 CF 直接 invisible 放行、不弹 interactive challenge。
 * 这是 SeleniumBase + uc_gui_click_captcha（在 Xvfb 里物理坐标点 Turnstile）被判可疑、
 * 拿不到 token 的死结所在。
 *
 * 流程：启动过盾浏览器（挂代理）→ 出口 IP 自检 → 打开登录页 → 关 cookie 弹窗 →
 *   填凭证 → 轮询 cf-turnstile-response token 非空为唯一权威信号（fail-closed，无 token 不点 Sign in）
 *   → 点 Sign in → 提交后再判一次是否又冒 Turnstile → 轮询跳转 /panel|/dashboard → 登录成功
 *   → 点 Extend → 点 Order now → 检查成功提示 → TG 通知 + 截图。
 *
 * 环境变量（沿用旧 Python 版，无需改 Secrets）：
 *   EMAIL            登录邮箱
 *   PASSWORD         登录密码
 *   IS_PROXY         "true" 时挂代理（默认走 socks5://127.0.0.1:1080）
 *   PROXY_SERVER     代理地址，默认 socks5://127.0.0.1:1080
 *   TG_BOT_TOKEN     Telegram bot token
 *   TG_CHAT_ID       Telegram chat id
 */

const { connect } = require('puppeteer-real-browser');

const EMAIL = process.env.EMAIL || '';
const PASSWORD = process.env.PASSWORD || '';
const IS_PROXY = (process.env.IS_PROXY || 'false').toLowerCase() === 'true';
const PROXY_SERVER = (process.env.PROXY_SERVER || '').trim() || 'socks5://127.0.0.1:1080';
const TG_BOT_TOKEN = process.env.TG_BOT_TOKEN || '';
const TG_CHAT_ID = process.env.TG_CHAT_ID || '';

const BASE_URL = 'https://client.therose.cloud/login';

// 掩码邮箱（只用于通知展示，脱敏）—— 与旧 Python 版 mask_email 等价
function maskEmail(email) {
    if (!email) return '（未配置）';
    if (email.includes('@')) {
        const [name, domain] = email.split('@', 2);
        if (name.length > 4) return `${name.slice(0, 2)}****${name.slice(-2)}@${domain}`;
        return `${name}@${domain}`;
    }
    return email.length > 2 ? email.slice(0, 2) + '****' : email + '****';
}

// HH:MM:SS → 秒，非法/空返回 0 —— 与 g4 timeToSeconds 等价
function timeToSeconds(t) {
    if (!t) return 0;
    const m = String(t).trim().match(/(\d{1,2}):(\d{2}):(\d{2})/);
    if (!m) return 0;
    return +m[1] * 3600 + +m[2] * 60 + +m[3];
}

function log(msg) {
    const t = new Date().toTimeString().slice(0, 8);
    console.log(`[${t}] [INFO] ${msg}`);
}

function sleep(ms) {
    return new Promise((r) => setTimeout(r, ms));
}

function humanWait(minS = 2, maxS = 4) {
    return sleep((minS + Math.random() * (maxS - minS)) * 1000);
}

// 当前 GMT+8 时间字符串，与旧 Python 版 format_notification 的执行时间一致
function nowBeijing() {
    const d = new Date();
    const beijing = new Date(d.getTime() + 8 * 3600 * 1000);
    const pad = (n) => String(n).padStart(2, '0');
    return `${beijing.getUTCFullYear()}-${pad(beijing.getUTCMonth() + 1)}-${pad(beijing.getUTCDate())} ${pad(beijing.getUTCHours())}:${pad(beijing.getUTCMinutes())}:${pad(beijing.getUTCSeconds())}`;
}

// 掩码出口 IP：只打点分首尾段，如 1.2.***.4
function maskIp(ip) {
    const p = String(ip || '').split('.');
    if (p.length === 4) return `${p[0]}.${p[1]}.***.${p[3]}`;
    return '未知';
}

// 发送 Telegram 通知（与旧 Python 版 send_tg 等价）
async function sendTelegram(message) {
    if (!TG_BOT_TOKEN || !TG_CHAT_ID) {
        log('⚠️ 未配置 TG_BOT_TOKEN / TG_CHAT_ID，跳过推送。');
        return;
    }
    try {
        const res = await fetch(`https://api.telegram.org/bot${TG_BOT_TOKEN}/sendMessage`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ chat_id: TG_CHAT_ID, text: message }),
        });
        if (res.ok) log('✅ TG 推送已发送');
        else log(`❌ TG 推送失败: HTTP ${res.status}`);
    } catch (e) {
        log(`❌ TG 推送异常: ${e.message}`);
    }
}

// 与旧 Python 版 format_notification 逐行一致的通知文案
function formatNotification(status, extra = '', error = '') {
    const lines = ['🌹 The Rose Cloud 续期通知', '', status, `👤 登录账户: ${maskEmail(EMAIL)}`];
    if (extra) lines.push(extra);
    if (error) lines.push(`⚠️ 错误信息: ${error}`);
    lines.push(`⏱️ 执行时间: ${nowBeijing()}`);
    return lines.join('\n');
}

// 启动过盾浏览器（puppeteer-real-browser + turnstile:true），按 IS_PROXY 挂 socks5 代理
async function launchRealBrowser() {
    const args = [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu',
        '--window-size=1280,1200',
    ];
    if (IS_PROXY) args.push(`--proxy-server=${PROXY_SERVER}`);

    log('🚀 启动浏览器（puppeteer-real-browser / turnstile）');
    let browser, page;
    try {
        ({ browser, page } = await connect({
            headless: false,
            turnstile: true,             // 自动求解 Cloudflare Turnstile（替代旧版 uc_gui_click_captcha）
            disableXvfb: true,            // 外层 workflow 已用 xvfb-run，避免嵌套 X server 冲突
            connectOption: { defaultViewport: null, executablePath: '/usr/bin/google-chrome' },
            args,
        }));
    } catch (e) {
        throw new Error(`浏览器启动失败: ${e.message}`);
    }
    await page.setViewport({ width: 1280, height: 1200 });
    return { browser, page };
}

// 读取 Turnstile token：先 window.turnstile.getResponse()，再隐藏字段兜底（与 g4 getTurnstileToken 等价）
async function getTurnstileToken(page) {
    try {
        return await page.evaluate(() => {
            try {
                if (window.turnstile && typeof window.turnstile.getResponse === 'function') {
                    const r = window.turnstile.getResponse();
                    if (r && r.length > 20) return r;
                }
            } catch (e) {}
            const el = document.querySelector('[name="cf-turnstile-response"]');
            return el && el.value && el.value.length > 20 ? el.value : '';
        });
    } catch (e) {
        return '';
    }
}

// 轮询等 Turnstile 被 puppeteer-real-browser 自动求解：唯一权威信号 = token 非空
async function waitTurnstileToken(page, timeoutS = 60) {
    log('📡 等待 puppeteer-real-browser 自动求解 Turnstile...');
    for (let i = 0; i < timeoutS; i++) {
        const token = await getTurnstileToken(page);
        if (token) {
            log(`✅ Turnstile token 已就绪（长度 ${token.length}）`);
            return true;
        }
        if (i === 20) log('⏳ Turnstile 仍在求解中（可能出现 interactive checkbox，自动求解器处理中）...');
        await sleep(1500);
    }
    return false;
}

// 关 cookie 同意弹窗（沿用 g4 evaluate 写法，消除遮挡）
async function dismissCookieConsent(page) {
    try {
        await page.evaluate(() => {
            const btns = Array.from(document.querySelectorAll('button, span, a'));
            const b = btns.find((el) => {
                const t = (el.textContent || '').trim().toLowerCase();
                return t === 'consent' || t === 'accept' || t === 'i agree' || t.includes('recommended cookies');
            });
            if (b) b.click();
        });
    } catch (e) { /* 忽略 */ }
}

// 填邮箱/密码，填后校验真写入、非 placeholder，异常重填 —— 对齐旧 Python 版 fill+校验逻辑
async function fillCredentials(page) {
    log('📧 填写邮箱...');
    // 先清空再填，避免残留
    await page.click('#login_form_email', { clickCount: 3 }).catch(() => {});
    await page.type('#login_form_email', EMAIL, { delay: 30 });
    log('🔑 填写密码...');
    await page.click('#login_form_password', { clickCount: 3 }).catch(() => {});
    await page.type('#login_form_password', PASSWORD, { delay: 30 });
    await sleep(400);
    const v = await page.evaluate(() => {
        const e = document.querySelector('#login_form_email');
        return e ? e.value || '' : '';
    });
    if (v && !v.includes(EMAIL.slice(0, 3)) && v !== EMAIL) {
        log(`⚠️ 邮箱字段异常（值前缀 ${v.slice(0, 6)}），重填`);
        await page.click('#login_form_email', { clickCount: 3 });
        await page.type('#login_form_email', EMAIL, { delay: 30 });
        await page.click('#login_form_password', { clickCount: 3 });
        await page.type('#login_form_password', PASSWORD, { delay: 30 });
    }
}

// 为 fail-closed 诊断采集登录页 DOM 状态
async function diagnosePage(page) {
    try {
        return await page.evaluate(() => {
            const errSel = ['.alert-danger', '.alert.alert-danger', 'div[role="alert"]', '.invalid-feedback', '.form-error', '.error-message', 'div.alert'];
            let errText = '';
            for (const s of errSel) {
                const el = document.querySelector(s);
                if (el && (el.offsetParent !== null || getComputedStyle(el).display !== 'none')) {
                    errText = (el.textContent || '').trim();
                    if (errText) break;
                }
            }
            const tEl = document.querySelector('[name="cf-turnstile-response"]');
            const cf = document.querySelector('iframe[src*="challenges.cloudflare.com"], iframe[src*="turnstile"], .cf-turnstile');
            return {
                url: location.href,
                title: document.title,
                hasCfIframe: !!cf,
                tokenLen: tEl && tEl.value ? tEl.value.length : 0,
                errText: errText.slice(0, 300),
            };
        });
    } catch (e) {
        return { diagError: e.message };
    }
}

// 登录流程：过盾 → fail-closed → 点 Sign in → 等跳 /panel|/dashboard
async function login(page) {
    log('🌐 打开登录页面...');
    await page.goto(BASE_URL, { waitUntil: 'domcontentloaded', timeout: 60000 });
    await humanWait(2, 4);
    // 若被 CF 拦到 challenge 中间页，puppeteer-real-browser 会自动求解；给一点缓冲
    await dismissCookieConsent(page);

    let tokenOk = false;
    for (let attempt = 1; attempt <= 3; attempt++) {
        // 每轮先确保凭证在（Turnstile 重新触发可能清表单）
        await fillCredentials(page);

        // puppeteer-real-browser turnstile:true 在后台自动求解，我们只等 token
        if (await waitTurnstileToken(page, 60)) {
            tokenOk = true;
            break;
        }
        log(`⏳ 第 ${attempt} 次未拿到 token，重试...`);
        // 触发重新渲染 Turnstile：整页重开
        if (attempt < 3) {
            await page.evaluate(() => { try { window.location.reload(); } catch (e) {} });
            await humanWait(4, 7);
            await dismissCookieConsent(page);
        }
    }

    if (!tokenOk) {
        const diag = await diagnosePage(page);
        log(`🩺 登录诊断: ${JSON.stringify(diag)}`);
        try { await page.screenshot({ path: 'artifacts/turnstile_fail.png' }); } catch (e) {}
        throw new Error('Cloudflare Turnstile 验证未通过：未拿到有效 token，已终止（不点 Sign in）。');
    }

    log('🖱️ 点击 Sign in...');
    try {
        await page.evaluate(() => {
            const btns = Array.from(document.querySelectorAll('button'));
            const b = btns.find((el) => (el.textContent || '').trim().toLowerCase() === 'sign in');
            if (b) b.click();
            else throw new Error('未找到 Sign in 按钮');
        });
    } catch (e) {
        // JS click 兜底
        await page.evaluate(() => {
            const b = document.querySelector('button');
            if (b) b.click();
        });
    }
    await humanWait(4, 7);

    // 提交后若又冒 Turnstile，再等一次 token
    const tokenAfter = await getTurnstileToken(page);
    if (!tokenAfter) {
        const cfStill = await page.evaluate(() => !!(document.querySelector('iframe[src*="challenges.cloudflare.com"], .cf-turnstile')));
        if (cfStill) {
            log('🛡 提交后再冒 Turnstile，继续等 token...');
            await waitTurnstileToken(page, 25);
            await sleep(1500);
        }
    }

    for (let i = 0; i < 30; i++) {
        const url = page.url();
        if (url.includes('/panel') || url.includes('/dashboard')) {
            log(`✅ 登录成功，已跳转: ${url}`);
            return url;
        }
        if (i > 0 && i % 5 === 0) {
            const diag = await diagnosePage(page);
            if (diag.errText && /invalid|incorrect|wrong|错误|失败/i.test(diag.errText)) {
                log(`❌ 检测到登录错误: ${diag.errText}`);
                try { await page.screenshot({ path: 'artifacts/login_error.png' }); } catch (e) {}
                throw new Error(`登录被拒: ${diag.errText || '未知错误'} | ${url}`);
            }
        }
        await sleep(1000);
    }
    const diag = await diagnosePage(page);
    try { await page.screenshot({ path: 'artifacts/login_timeout.png' }); } catch (e) {}
    throw new Error(`登录超时未跳转 Dashboard | ${diag.url || ''} | ${diag.errText || ''}`);
}

// 纯逻辑导出，供 tests/ 断言
module.exports = { maskEmail, timeToSeconds };
